components/sup/src/main.rs

```diff
@@ -202,9 +203,17 @@ async fn sub_run_rsr_imlw_mlw_gsw_smw_rhw_msw(sup_run: SupRun,
                                               -> Result<()> {
     set_supervisor_logging_options(&sup_run);
 
+    let mut svc_load_msgs =
+        svc::svc_loads_from_paths(&sup_run.svc_config_paths)?.into_iter()
+                                                             .map(|svc_load| svc_load.into())
+                                                             .collect::<Vec<_>>();
```

Comment:
This will work nicely the first time a supervisor is run on a machine, but every time thereafter, I think we're going to run into potential issues. Once a service has been loaded, a spec file will exist, which is what the Supervisor reads to know which services it's responsible for. Those spec files may also differ from what's in these config files (e.g., I start up a service based on a config file, then change how its configured using `hab svc load --force`).

Additionally, if you `hab svc unload`ed a previously-loaded service, but didn't also remember to remove the config file, that service would come back the next time you started the Supervisor.

I believe all this would also be in play when the Supervisor restarted after an update or `SIGHUP`, too.

I worry about the potential for conflicts, "spooky action", and impacts to running services. Having two different files to represent how a given service is supposed to run leaves a lot of space for weird behavior. Ultimately, I think I'd like to come up with a solution that unifies this config file with the spec file, formally exposes the spec file as "the way" to manage services, or some other approach that leaves us with a single source of truth for this information.

Reply:
Thanks! These are very useful cases to consider.

> This will work nicely the first time a supervisor is run on a machine, but every time thereafter, I think we're going to run into potential issues.

As long as the `force` option is set, I believe it will "just work".

> Once a service has been loaded, a spec file will exist, which is what the Supervisor reads to know which services it's responsible for. Those spec files may also differ from what's in these config files (e.g., I start up a service based on a config file, then change how its configured using hab svc load --force).

True. In this case we will revert back to what is in the config file on restart. I think the answer might be simply "don't do that" :sweat_smile:. If you are going to use config files you should probably always use config files and run `hab svc bulkload" to apply new changes instead of `hab svc load --force`.

> Additionally, if you hab svc unloaded a previously-loaded service, but didn't also remember to remove the config file, that service would come back the next time you started the Supervisor.

This is a tough one. Maybe we could have an option on Supervisor startup to unload any services that are running and were not explicitly found in the configuration?

> I believe all this would also be in play when the Supervisor restarted after an update or SIGHUP, too.

I think this just works as long as the `force` option is set.

I think of the config files as duplicating the functionality of `hab svc load` and not duplicating the functionality of spec files. The `hab svc load` and service config files act as two APIs for instructing the Supervisor what data to write to the spec file. The reason for two APIs existing is because there are two separate modes in which you want to interact with the Supervisor. One is very interactive and iterative (ie the cli). One is set it once and forget it (ie the service config files). I agree you can get a lot of odd behavior mixing these two modes, but maybe we can simply say "dont do that". I am certainly open to a better answer, but I am not sure what it is yet.

Reply:
To the extent possible, I'd like to avoid having a lot of "don't do that" functionality. It's a bad user experience to have something fail or behave contrary to what you're expecting and have the response be "don't do that" (a rational response to that would be "well, why did you _let_ me do that?"). If we know that something is going to be potentially confusing in some way, we should build it in a way that's not confusing, or anticipate potential conflicts and make them impossible.

For instance, in this case, one potential avenue might be to have the Supervisor know which mode its supposed to be operating in ("config file mode" or "spec file mode") and behave appropriately. This could extend to having `hab svc load --force` calls fail with an explanatory message if the Supervisor is in "config file mode". (This is just an example for argument's sake; I'm skeptical that this concrete design would be an overall win from a UX perspective.)